cedar-policy-validator/src/entity_manifest.rs

```diff
@@ -0,0 +1,1266 @@
+// Don't make fields `pub`, don't make breaking changes, and use caution
+// when adding public methods.
+#[doc = include_str!("../../cedar-policy/experimental_warning.md")]
+#[serde_as]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct RootAccessTrie<T = ()>
+where
+    T: Clone,
+{
+    /// The data that needs to be loaded, organized by root.
+    #[serde_as(as = "Vec<(_, _)>")]
+    #[serde(bound(deserialize = "T: Default"))]
+    pub trie: HashMap<EntityRoot, AccessTrie<T>>,
+}
+
+/// A Trie representing a set of data paths to load,
+/// starting implicitly from a Cedar value.
+///
+/// `T` represents an optional type annotation on each
+/// node in the [`AccessTrie`].
+///
+// CAUTION: this type is publicly exported in `cedar-policy`.
+// Don't make fields `pub`, don't make breaking changes, and use caution
+// when adding public methods.
+#[doc = include_str!("../../cedar-policy/experimental_warning.md")]
+#[serde_as]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct AccessTrie<T = ()> {
+    /// Child data of this entity slice.
+    /// The keys are edges in the trie pointing to sub-trie values.
+    #[serde_as(as = "Vec<(_, _)>")]
+    pub children: Fields<T>,
+    /// For entity types, this boolean may be `true`
+    /// to signal that all the ancestors in the entity hierarchy
+    /// are required (transitively).
+    pub ancestors_required: bool,
+    /// Optional data annotation, usually used for type information.
+    #[serde(skip_serializing, skip_deserializing)]
+    #[serde(bound(deserialize = "T: Default"))]
+    pub data: T,
+}
+
+/// A data path that may end with requesting the parents of
+/// an entity.
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct AccessPath {
+    /// The root variable that begins the data path
+    pub root: EntityRoot,
+    /// The path of fields of entities or structs
+    pub path: Vec<SmolStr>,
+    /// Request all the parents in the entity hierarchy of this entity.
+    pub ancestors_required: bool,
+}
+
+/// Entity manifest computation does not handle the full
+/// cedar language. In particular, the policies must follow the
+/// following grammar:
+/// ```text
+/// <expr> = <datapath-expr>
+///          <datapath-expr> in <expr>
+///          <expr> + <expr>
+///          if <expr> { <expr> } { <expr> }
+///          ... all other cedar operators not mentioned by datapath-expr
+
+/// <datapath-expr> = <datapath-expr>.<field>
+///                   <datapath-expr> has <field>
+///                   <variable>
+///                   <entity literal>
+/// ```
+/// The `get_expr_path` function handles `datapath-expr` expressions.
+/// This error message tells the user not to use certain operators
+/// before accessing record or entity attributes, breaking this grammar.
+#[derive(Debug, Clone, Error, Hash, Eq, PartialEq)]
+#[error("For policy `{policy_id}`, failed to analyze expression while computing entity manifest.`")]
```

Comment:
IIRC. our error message convention is to omit trailing punctuation and not capitalize the first letter.